File: java/examples/src/main/java/com/optikpi/examples/TestGamingEndpoint.java
package com.optikpi.examples;

import java.math.BigDecimal;
import java.time.Instant;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.optikpi.datapipeline.ClientConfig;
import com.optikpi.datapipeline.OptikpiDataPipelineSDK;
import com.optikpi.datapipeline.model.GamingActivityEvent;
import com.optikpi.datapipeline.model.ValidationResult;

import io.github.cdimascio.dotenv.Dotenv;

public class TestGamingEndpoint {

    private static final ObjectMapper mapper = new ObjectMapper()
            .enable(SerializationFeature.INDENT_OUTPUT);
    public static void main(String[] args) {
        Dotenv dotenv = Dotenv.configure()
                .directory(".")
                .filename(".env")
                .load();

        String authToken = dotenv.get("AUTH_TOKEN");
        String accountId = dotenv.get("ACCOUNT_ID");
        String workspaceId = dotenv.get("WORKSPACE_ID");
        String baseUrl = dotenv.get("API_BASE_URL");

        if (authToken == null || accountId == null || workspaceId == null) {
            System.err.println("Error: Missing required environment variables");
            System.exit(1);
        }

        ClientConfig config = new ClientConfig(authToken, accountId, workspaceId);
        config.setBaseUrl(baseUrl);

        OptikpiDataPipelineSDK sdk = new OptikpiDataPipelineSDK(config);

        System.out.println("🚀 Testing Gaming Event Endpoints");
        System.out.println("==================================");
        System.out.println("Configuration:");
        System.out.println("📌 API Base URL: " +  config.getBaseUrl());
        System.out.println("👤 Account ID: " + config.getAccountId());
        System.out.println("🏢 Workspace ID: " + config.getWorkspaceId());
        System.out.println();
        
        testGamingEvent(sdk, accountId, workspaceId);
    }

    private static void testGamingEvent(OptikpiDataPipelineSDK sdk, String accountId, String workspaceId) {
        try {
            GamingActivityEvent event = new GamingActivityEvent();
            event.setAccountId(accountId);
            event.setWorkspaceId(workspaceId);
            event.setUserId("java_012");
            event.setEventCategory("Gaming Activity");
            event.setEventName("Play Casino Game");
            event.setEventId("evt_" + System.currentTimeMillis());
            event.setEventTime(Instant.now().toString());
            event.setWagerAmount(new BigDecimal("10.00"));
            event.setWinAmount(new BigDecimal("25.00"));
            event.setLossAmount(new BigDecimal("0.00"));
            event.setGameId("game_001");
            event.setGameTitle("Mega Fortune Slots");
            event.setProvider("ProviderXYZ");
            event.setBonusId("bonus_12345");
            event.setFreeSpinId("freespin_67890");
            event.setJackpotAmount(new BigDecimal("1000.00"));
            event.setNumSpinsPlayed(50);
            event.setGameTheme("Egyptian");
            event.setRemainingSpins(10);
            event.setBetValuePerSpin(new BigDecimal("0.50"));
            event.setWageringRequirementsMet(true);
            event.setFreeSpinExpiryDate("2024-12-31T23:59:59Z");
            event.setCampaignId("camp_summer2024");
            event.setCampaignName("Summer Bonanza");
            event.setRtp(new BigDecimal("96.5"));
            event.setGameCategory("slots");
            event.setWinningBetAmount(new BigDecimal("25.00"));
            event.setJackpotType("progressive");
            event.setVolatility("high");
            event.setMinBet(new BigDecimal("0.10"));
            event.setMaxBet(new BigDecimal("100.00"));
            event.setNumberOfReels(5);
            event.setNumberOfPaylines(20);
            event.setFeatureTypes("wild,scatter,freespins");
            event.setGameReleaseDate("2023-01-15T00:00:00Z");
            event.setLiveDealerAvailability(false);
            event.setSideBetsAvailability(true);
            event.setMultiplayerOption(false);
            event.setAutoPlay(true);
            event.setPokerVariant("texas_holdem");
            event.setTournamentName("Weekend Warriors");
            event.setBuyInAmount(new BigDecimal("50.00"));
            event.setTableType("cash_game");
            event.setStakesLevel("medium");
            event.setNumberOfPlayers(6);
            event.setGameDuration(45);
            event.setHandVolume(120);
            event.setPlayerPosition("button");
            event.setFinalHand("royal_flush");
            event.setRakeContribution(new BigDecimal("2.50"));
            event.setMultiTablingIndicator(false);
            event.setSessionResult("win");
            event.setVipStatus("gold");
            event.setBlindLevel("50/100");
            event.setRebuyAndAddonInfo("1 rebuy, 1 addon");
            event.setSportType("football");
            event.setBettingMarket("match_winner");
            event.setOdds(new BigDecimal("2.50"));
            event.setLiveBettingAvailability(true);
            event.setResult("won");
            event.setBetStatus("settled");
            event.setBettingChannel("online");
            event.setBonusType("welcome_bonus");
            event.setBonusAmount(new BigDecimal("100.00"));
            event.setFreeSpinStartDate("2024-01-01T00:00:00Z");
            event.setNumSpinsAwarded(20);
            event.setBonusCode("WELCOME100");
            event.setParentGameCategory("casino");
            event.setCurrency("USD");
            event.setMoneyType("real");
            event.setTransactionType("bet");
            
            ValidationResult valid = event.validate();
            if (!valid.isValid()) {
                System.out.println("\n❌ Validation Failed!");
                System.out.println("Errors: " + valid.getErrors());
                return;
            }
            
            System.out.println("✅ Gaming Activity event validated successfully!");
            System.out.println("\n📋Gaming Activity Event Data:");
            System.out.println(mapper.writeValueAsString(event));

            System.out.println("\n🕒 making API request using SDK...");
            long start = System.currentTimeMillis();
            var response = sdk.sendGamingActivityEvent(event);
            long end = System.currentTimeMillis();

            System.out.println("\n📡 API Response:");
            System.out.println("⏱ Response Time: " + (end - start) + "ms");
            System.out.println("HTTP Status: " + response.getStatus());
            System.out.println("Sending Gaming event...");
            
            if (response.isSuccess()) {
                System.out.println("✅ SUCCESS!");
                System.out.println("Response: " +
                        mapper.writeValueAsString(response.getData()));
            } else {
                System.out.println("❌ FAILED!");
                System.out.println("Error: " +
                        mapper.writeValueAsString(response.getError()));
            }

        } catch (Exception e) {
            System.err.println("\n💥 Exception occurred:");
            e.printStackTrace();
        }
    }
}

File: php/examples/test-gaming-endpoint.php
<?php

require_once __DIR__ . '/../vendor/autoload.php';

use Optikpi\DataPipeline\OptikpiDataPipelineSDK;
use Optikpi\DataPipeline\Models\GamingActivityEvent;

// Load environment variables
$dotenv = parse_ini_file(__DIR__ . '/../.env');

$API_BASE_URL = $dotenv['API_BASE_URL'] ?? getenv('API_BASE_URL');
$AUTH_TOKEN = $dotenv['AUTH_TOKEN'] ?? getenv('AUTH_TOKEN');
$ACCOUNT_ID = $dotenv['ACCOUNT_ID'] ?? getenv('ACCOUNT_ID');
$WORKSPACE_ID = $dotenv['WORKSPACE_ID'] ?? getenv('WORKSPACE_ID');

// Validate required environment variables
if (empty($AUTH_TOKEN) || empty($ACCOUNT_ID) || empty($WORKSPACE_ID)) {
    echo "❌ Error: Missing required environment variables!\n";
    echo "   Please set: AUTH_TOKEN, ACCOUNT_ID, WORKSPACE_ID\n";
    echo "   Copy env.example to .env and fill in your values\n";
    exit(1);
}

// Initialize SDK
$sdk = new OptikpiDataPipelineSDK([
    'authToken' => $AUTH_TOKEN,
    'accountId' => $ACCOUNT_ID,
    'workspaceId' => $WORKSPACE_ID,
    'baseURL' => $API_BASE_URL
]);

// Gaming activity event data
$gaming = new GamingActivityEvent([
    'account_id' => $ACCOUNT_ID,
    'workspace_id' => $WORKSPACE_ID,
    'user_id' => 'user123411',
    'event_category' => 'Gaming Activity',
    'event_name' => 'Play Casino Game',
    'event_id' => 'evt_' . (time() * 1000),
    'event_time' => date('c'),
    'wager_amount' => 10.00,
    'win_amount' => 25.00,
    'loss_amount' => 0.00,
    'game_id' => 'game_001',
    'game_title' => 'Mega Fortune Slots',
    'provider' => 'ProviderXYZ',
    'bonus_id' => 'bonus_12345',
    'free_spin_id' => 'freespin_67890',
    'num_spins_played' => 50,
    'game_theme' => 'Egyptian',
    'remaining_spins' => 10,
    'bet_value_per_spin' => 0.50,
    'wagering_requirements_met' => true,
    'free_spin_expiry_date' => '2024-12-31T23:59:59Z',
    'campaign_id' => 'camp_summer2024',
    'campaign_name' => 'Summer Bonanza',
    'rtp' => 96.5,
    'game_category' => 'slots',
    'winning_bet_amount' => 25.00,
    'jackpot_type' => 'progressive',
    'volatility' => 'high',
    'min_bet' => 0.10,
    'max_bet' => 100.00,
    'number_of_reels' => 5,
    'number_of_paylines' => 20,
    'feature_types' => 'wild,scatter,freespins',
    'game_release_date' => '2023-01-15T00:00:00Z',
    'live_dealer_availability' => false,
    'side_bets_availability' => true,
    'multiplayer_option' => false,
    'auto_play' => true,
    'poker_variant' => 'texas_holdem',
    'tournament_name' => 'Weekend Warriors',
    'buy_in_amount' => 50.00,
    'table_type' => 'cash_game',
    'stakes_level' => 'medium',
    'number_of_players' => 6,
    'game_duration' => 45,
    'hand_volume' => 120,
    'player_position' => 'button',
    'final_hand' => 'royal_flush',
    'rake_contribution' => 2.50,
    'multi_tabling_indicator' => false,
    'session_result' => 'win',
    'vip_status' => 'gold',
    'blind_level' => '50/100',
    'rebuy_and_addon_info' => '1 rebuy, 1 addon',
    'sport_type' => 'football',
    'betting_market' => 'match_winner',
    'odds' => 2.50,
    'live_betting_availability' => true,
    'result' => 'won',
    'bet_status' => 'settled',
    'betting_channel' => 'online',
    'bonus_type' => 'welcome_bonus',
    'bonus_amount' => 100.00,
    'free_spin_start_date' => '2024-01-01T00:00:00Z',
    'num_spins_awarded' => 20,
    'bonus_code' => 'WELCOME100',
    'parent_game_category' => 'casino',
    'currency' => 'USD',
    'money_type' => 'real',
    'transaction_type' => 'bet'
]);

// Validate the gaming event
$validation = $gaming->validate();
if (!$validation['isValid']) {
    echo "❌ Validation errors:\n";
    foreach ($validation['errors'] as $error) {
        echo "  - $error\n";
    }
    exit(1);
}

echo "✅ Gaming event validated successfully!\n";

// SDK handles HMAC authentication automatically

// Test gaming endpoint
function testGamingEndpoint($sdk, $gaming, $API_BASE_URL, $ACCOUNT_ID, $WORKSPACE_ID, $AUTH_TOKEN) {
    try {
        echo "🎮 Testing Gaming Activity Events Endpoint\n";
        echo "==========================================\n";

        echo "Configuration:\n";
        echo "API Base URL: $API_BASE_URL\n";
        echo "Account ID: $ACCOUNT_ID\n";
        echo "Workspace ID: $WORKSPACE_ID\n";
        echo "Auth Token: " . substr($AUTH_TOKEN, 0, 8) . "...\n";

        echo "\nMaking API request using SDK...\n";
        echo "Gaming Event Data: " . json_encode($gaming->toArray(), JSON_PRETTY_PRINT) . "\n";

        // Make the API call using SDK
        $startTime = microtime(true) * 1000;
        $result = $sdk->sendGamingActivityEvent($gaming);
        $endTime = microtime(true) * 1000;

        if ($result['success']) {
            echo "\n✅ Success!\n";
            echo "==========================================\n";
            echo "HTTP Status: " . $result['status'] . "\n";
            echo "Response Time: " . round($endTime - $startTime) . "ms\n";
            echo "SDK Success: " . ($result['success'] ? 'true' : 'false') . "\n";
            echo "Response Data: " . json_encode($result['data'], JSON_PRETTY_PRINT) . "\n";
        } else {
            echo "\n❌ API Error!\n";
            echo "==========================================\n";
            echo "HTTP Status: " . ($result['status'] ?? 'N/A') . "\n";
            echo "Response Time: " . round($endTime - $startTime) . "ms\n";
            echo "SDK Success: " . ($result['success'] ? 'true' : 'false') . "\n";
            echo "Error Data: " . json_encode($result['data'], JSON_PRETTY_PRINT) . "\n";
        }

    } catch (Exception $error) {
        echo "\n❌ SDK Error occurred!\n";
        echo "==========================================\n";
        echo "Error: " . $error->getMessage() . "\n";
        echo "Stack: " . $error->getTraceAsString() . "\n";
    }
}

// Run the test
if (php_sapi_name() === 'cli') {
    testGamingEndpoint($sdk, $gaming, $API_BASE_URL, $ACCOUNT_ID, $WORKSPACE_ID, $AUTH_TOKEN);
}
